Add adaptive replacement cache

qwe661234 opened the pull request:
Current basic block management consumes a significant amount of memory, which leads to unnecessary waste due to frequent map allocation and release. Adaptive Replacement Cache (ARC) is a page replacement algorithm with better performance than least recently used (LRU). After the translated blocks are handled by ARC, better memory usage and hit rates can be achieved by keeping track of frequently used and recently used pages, as well as a recent eviction history for both.

According to the cache information obtained while running CoreMark, the cache hit rate of ARC can reach over 99%.

see #105

jserv commented on Makefile:
```diff
@@ -74,6 +74,14 @@ gdbstub-test: $(BIN)
 endif
 
+# Import adaptive replacement cache
+ENABLE_ARCACHE ?= 1
```
Don't add both `ENABLE_ARCACHE` and `ENABLE_ARCACHE_INFO`.
I would like to have a minimal ARC implementation along with a test program. Then, we can integrate the refined one into rv32emu core in further pull requests.

jserv replied:
There should no entry `ENABLE_ARCACHE`. That is, this pull patch only delivers ARC header and C implementation along with test program(s). You should tweak `Makefile` for the linkage of ARC test program(s).

qwe661234 replied:
OK, I understand.

jserv replied:
The name `ARCACHE` sounds stupid since ARC stands for Adaptive Replacement Cache.

jserv replied:
This ARC implementation is meant be reusable for further JIT compilation. Its quality should be verified, and it is exactly why test programs should appear for validation purposes. The interface to ARC should be generic and straightforward without disclosing the details.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,230 @@
+#include "cache.h"
```
Add copyright notice.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,230 @@
+#include "cache.h"
+
+#define min(a, b) ((a < b) ? a : b)
```
Capitalize these macros. That is, `MIN` and `MAX`.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,42 @@
+    B1: LRU Ghost List
+    B2: LFU Ghost List
+*/
+typedef enum { T1, B1, T2, B2 } arc_type_t;
```
Don't call it `arc_type_t`.
The implementation details should appear in C source rather than public headers.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,42 @@
+    struct list_head *list_table[4];
+    uint32_t list_size[4];
+    struct hashtable *map;
+    uint32_t c;
```
It is hard to know what `c` and `p` are.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,42 @@
+    struct hashtable *map;
+    uint32_t c;
+    uint32_t p;
+#if RV32_HAS(ARCACHE_INFO)
```
Don't add `ARCACHE_INFO`. Minimize the code as possible.

qwe661234 replied:
This flag is for showing cache information, such as cache hit rate and miss rate.

jserv commented on Makefile:
```diff
@@ -4,7 +4,7 @@ include mk/toolchain.mk
 BIN := $(OUT)/rv32emu
 
-CFLAGS = -std=gnu99 -O2 -Wall -Wextra
+CFLAGS = -std=gnu99 -O2 -Wall -Wextra -lpthread
```
Why pthread?

jserv commented on Makefile:
```diff
@@ -55,7 +55,7 @@ $(OUT)/syscall_sdl.o: CFLAGS += $(shell sdl2-config --cflags)
 endif
 
-ENABLE_GDBSTUB ?= 1
+ENABLE_GDBSTUB ?= 0
```
Don't change this.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,241 @@
+        INIT_LIST_HEAD(cache->list_table[i]);
+        cache->list_size[i] = 0;
+    }
+    cache->map = (hashtable_t *) malloc(sizeof(hashtable_t));
```
Shorten the code.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,241 @@
+        replace_LRU_list(cache);
+        move_to_mru(cache, entry, LFU_list);
+    }
+#if CACHE_INFO
```
Remove all occurrences of `CACHE_INFO`.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,44 @@
+
+struct hashtable;
+
+typedef struct cache {
```
Make use of forward declarations, so that we don't have to disclose the implementation details.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,44 @@
+#endif
+} cache_t;
+
+cache_t *cache_create();
```
You MUST write down comments for each public function.

jserv commented on src/emulate.c:
```diff
@@ -1250,7 +1250,7 @@ static bool insn_is_branch(uint8_t opcode)
     return false;
 }
-
+#if !RV32_HAS(ARCACHE)
```
Don't integrate `ARCACHE` at the moment.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,44 @@
+#include <stdlib.h>
+#include <string.h>
+
+#include "list.h"
```
Don't expose `list.h` in this public header. Encapsulate the details as possible.

jserv replied:
Then, it is feasible to switch among different cache replacement implementations later. Let's stick to consistent interface.

jserv commented on src/list.h:
```diff
@@ -0,0 +1,431 @@
+/* Linux-like doubly-linked list implementation */
```
Shrink this header to keep necessary macros and functions.

jserv commented:
Check the build system of [linux-list](https://github.com/sysprog21/linux-list) for building test programs.

qwe661234 commented:
> Check the build system of [linux-list](https://github.com/sysprog21/linux-list) for building test programs.

Does the test programs here means new test case for testing cache?

jserv commented:
> Does the test programs here means new test case for testing cache?

Exactly. These test programs should be built and verified before we integrate ARC into rv32emu core.

jserv reviewed (changes requested):
1. All test cases should be built from top-level Makefile.
2. Test specific file should be placed in tests directory, and there are no file names in upper case.
3. Show report of test cases and integrate into GitHub Actions.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+    return delete_value;
+}
+
+void cache_print_stats(cache_t *cache)
```
It is not necessary to print.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,46 @@
+#include <stdlib.h>
+#include <string.h>
+
+typedef enum { LRU_list, LFU_list, LRU_ghost_list, LFU_ghost_list } list_type_t;
```
Avoid disclosing implementation details. This cache interface should have nothing associated with ARC.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,46 @@
+
+#pragma once
+
+#include <assert.h>
```
Clean up inclusions. Only keep the headers you need for the declarations.

jserv commented on src/list.h:
```diff
@@ -0,0 +1,241 @@
+    next->prev = prev;
+    prev->next = next;
+
+#ifdef LIST_POISONING
```
Remove `LIST_POISONING` since we don't tend to utilize it.

jserv commented on tests/cache-test/src/testCache.c:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+int main(__attribute__((unused)) int argc, char *argv[])
```
Replace `_attribute__((unused))` with `UNUSED` since `common.h` already defines.

jserv replied:
You shall check `argc` in advance.

jserv commented on tests/cache-test/Makefile:
```diff
@@ -0,0 +1,52 @@
+$(BUILDDIR)/cache.o: $(SOURCE)/cache.c
+	$(Q)$(CC) -o $@ $(CFLAGS) -c -MMD -MF $@.d $<
+
+$(BUILDDIR)/testCache.o: src/testCache.c
```
Shorten the file names. Let's specify `tests/cache.c` for short.

jserv commented on tests/cache-test/Makefile:
```diff
@@ -0,0 +1,52 @@
+OBJS := $(addprefix $(BUILDDIR)/, $(OBJS))
+deps := $(OBJS:%.o=%.o.d)
+
+CHECK_ELF_FILES := \
```
You may consider to check shecc for its test suite.
See [tests/driver.sh](https://github.com/jserv/shecc/blob/master/tests/driver.sh) and [Makefile](https://github.com/jserv/shecc/blob/master/Makefile).

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+ * 4. size of LRU_list + size of LFU_list + size of LRU_ghost_list + size of
+ * LFU_ghost_list <= 2c
+ */
+void assert_cache(cache_t *cache)
```
It tends to be a private function. In addition, use macro to rework.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+    struct hashtable *map;
+    uint32_t capacity;
+    uint32_t lru_capacity;
+    uint64_t get_time;
```
Check if `get_time` and `hit_time` are necessary.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+
+typedef struct cache {
+    struct list_head *list_table[4];
+    uint32_t list_size[4];
```
Reduce the memory usage.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+#define BITS 8
+#define SIZE 256
+#define GOLDEN_RATIO_32 0x61C88647
+#define HASH(val) (((val) *GOLDEN_RATIO_32) >> (32 - BITS)) % SIZE
```
Use an inline function rather than a macro for hash function.

jserv replied:
Rewrite as following:
```c
#define HASH(val) (((val) * (GOLDEN_RATIO_32)) >> (32 - BITS)) % SIZE
```

Therefore, clang-format can be happier.

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,46 @@
+/** cache_create - crate a new cache
+ * @return: a pointer points to new cache
+ */
+struct cache *cache_create();
```
How can we specify the capacity?

qwe661234 replied:
I make the user specify the capacity, but this implementation significantly reduces performance.

jserv replied:
Why?

qwe661234 replied:
I guess the reason is that the hash function, which calculating by cache size, is called frequently. So, if we use a macro to represent size, the compiler can optimize it, possibly through constant propagation.

jserv replied:
Good point. Clarify your assumptions.

qwe661234 replied:
I examine the assembly code on x86 architecture, if we use a macro to represent size, the hash function only needs 5 instructions. But if we use a   variable to represent size, the hash function needs 9 instructions, including the dividing instruction which has large overhead.
* MACRO
```=
imul   $0x61c88647,%esi,%eax
mov    0x30(%rcx),%rdx
mov    (%rdx),%rdx
shr    $0x18,%eax
mov    (%rdx,%rax,8),%rax
```
* variable 
```=
imul   $0x61c88647,%esi,%eax
mov    $0x20,%ecx
xor    %edx,%edx
mov    0x30(%rdi),%r9
sub    0x6a84(%rip),%ecx        # 1204c <cache_size_bits>
shr    %cl,%eax
divl   0x6a80(%rip)        # 12050 <cache_size>
mov    (%r9),%rax
mov    (%rax,%rdx,8),%rax
```

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+
+#define MIN(a, b) ((a < b) ? a : b)
+#define BITS 8
+#define SIZE 256
```
Change the name `SIZE` because it is too generic.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+ */
+
+#include "cache.h"
+#include "list.h"
```
At present, header `list.h` is only used by `src/cache.c`, and we can apply the following changes to minimize overall code size:
1. Move the definition of `container_of` into `src/common.h`
2. Move the macros and inline functions into `src/cache.c`

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+           2 * cache->capacity);
+}
+
+void move_to_mru(cache_t *cache, arc_entry_t *entry, const list_type_t arc_type)
```
Do not expose the internal functions. That is, add `static`.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+        return NULL;
+
+    arc_entry_t *entry = NULL;
+    list_for_each_entry(entry, &cache->map->ht_list_head[HASH(key)], ht_list)
```
Update `.clang-format` to identify for-each macros. See https://github.com/sysprog21/lab0-c/blob/master/.clang-format

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+    uint32_t key;
+    list_type_t arc_type;
+    struct list_head list;
+    struct list_head ht_list;
```
Woud you consider the use of `hlist` used in Linux kernel?

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+
+typedef struct hashtable {
+    struct list_head *ht_list_head;
+} hashtable_t;
```
Ditto. You may learn from the hash table implementation in Linux kernel.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+
+cache_t *cache_create()
+{
+    cache_t *cache = malloc(sizeof(cache_t));
```
Validate the return value of `malloc`.

jserv commented on tests/cache-test/src/testCache.c:
```diff
@@ -0,0 +1,64 @@
+#include "cache.h"
+#define TESTSIZE 1
+
+void printValue(int *val)
```
Enforce consistent style.

jserv reviewed (changes requested):
All files in directory `tests` should comply with the following naming scheme.
> `tests/cache/ITEM.TYPE`

* `ITEM` is the test item. e.g., `cache_get`
* `TYPE` is the type for given tests. e.g., `in` for input commands; `expect` for expected results, `out` for something generated .

Therefore, we would have `tests/cache/cache_get.in` .

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,369 @@
+} list_type_t;
+
+struct list_head {
+    struct list_head *prev;
```
Rewrite as following:
```c
    struct list_head *prev, *next;
```

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,369 @@
+    LRU_ghost_list,
+    LFU_ghost_list,
+    list_quantity
+} list_type_t;
```
`list_type_t` is a bit confusing. Someone might regard it as the types of linked list.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,369 @@
+    uint32_t lru_capacity;
+} cache_t;
+
+#if defined(__GNUC__)
```
You shall provide the generic code without `typeof`.

jserv replied:
In addition, check clang as well. That is,
```c
#if defined(__GNUC__) || defined(__clang__)
```

jserv replied:
Don't touch `__LIST_HAVE_TYPEOF` since `src/common.h` already takes care of it.

jserv commented on src/common.h:
```diff
@@ -49,3 +49,30 @@
 /* run the 1st parameter */
 #define IIF_1(t, ...) t
+
+#if defined(__GNUC__)
```
Consolidate the checking for `typeof`.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,369 @@
+         &entry->member != (head); entry = safe,                           \
+        safe = list_entry(safe->member.next, __typeof__(*entry), member))
+
+#undef __LIST_HAVE_TYPEOF
```
Don't touch `__LIST_HAVE_TYPEOF`. Ensure that List API works for non-gcc/clang compilers.

jserv commented on mk/cache-test.mk:
```diff
@@ -0,0 +1,42 @@
+CACHE_TEST_DIR := tests/cache
+CACHE_BUILD_DIR := build/cache
+TARGET := test_cache.out
+SOURCE := src
```
Is `SOURCE` used?

jserv commented on mk/cache-test.mk:
```diff
@@ -0,0 +1,42 @@
+cache :$(TARGET)
+	$(Q)$(foreach e,$(CACHE_CHECK_ELF_FILES),\
+	    $(PRINTF) "Running $(e) ... "; \
+	    if [ "$(strip $(shell cat $(CACHE_TEST_DIR)/$(e).expect))" = "$(strip $(shell ./$(CACHE_BUILD_DIR)/$(TARGET) $(CACHE_TEST_DIR)/$(e).in))" ]; then \
```
Shouldn't you compare `.out` (the output) and `.expect` (the expected result)?

qwe661234 replied:
I campare the output (generated by printf) and .expect directly.

jserv replied:
How about using [cmp(1)](https://man7.org/linux/man-pages/man1/cmp.1.html)?

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,369 @@
+typedef struct cache {
+    struct list_head *list_table[list_quantity];
+    uint32_t list_size[list_quantity];
+    struct hashtable *map;
```
Enforce the use of `hashtable_t`.

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,369 @@
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) % (cache_size)
+
+int cache_size, cache_size_bits;
```
Both `cache_size` and `cache_size_bits` should be private, meaning that you shall mark them as `static`.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,231 @@
+
+void cache_free(cache_t *cache, void (*release_entry)(void *))
+{
+    for (int i = 0; i < 4; i++) {
```
Avoid hardcoded `4`.

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,378 @@
+    LRU_ghost_list,
+    LFU_ghost_list,
+    list_quantity
+} list_category_t;
```
Rename to `cache_list_t` and some comments like
> ARC improves the basic LRU strategy by splitting the cache directory into two lists, T1 and T2, for recently and frequently referenced entries. In turn, each of these is extended with a ghost list (B1 or B2), which is attached to the bottom of the two lists.

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,378 @@
+    LFU_list,
+    LRU_ghost_list,
+    LFU_ghost_list,
+    list_quantity
```
Change the name `list_quantity` to `N_CACHE_LIST_TYPES` for better understanding.

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,378 @@
+    struct hlist_node ht_list;
+} arc_entry_t;
+
+typedef struct hashtable {
```
Replace `struct hashtable` with `struct` since you are using typedef.

jserv commented on src/common.h:
```diff
@@ -49,3 +49,28 @@
 #define IIF_1(t, ...) t
+
+#if defined(__GNUC__) || defined(__clang__)
+#define __LIST_HAVE_TYPEOF 1
```
Rename to `__HAVE_TYPEOF`.

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,369 @@
+ * 4. size of LRU_list + size of LFU_list + size of LRU_ghost_list + size of
+ * LFU_ghost_list <= 2c
+ */
+#define ASSERT_CACHE(cache)                                                 \
```
Rename to `CACHE_ASSERT`.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,378 @@
+
+typedef struct cache {
+    struct list_head *list_table[list_quantity];
+    uint32_t list_size[list_quantity];
```
What are typical ranges for the lists used in ARC? I don't think it is worthy to preserve them in `uint32_t`.

qwe661234 replied:
The maximum of list size T1, T2, and B1 is equal to `cache_size / 2`, and the maximum of list size B2 is equal to `cache_size`. Therefore, if user set `cache_size` to more than 2^17, we would need `uint32_t` to store list_size. Otherwise, it might cause overflow.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,378 @@
+    if (!cache->map->ht_list_head) {
+        free(cache);
+        free(cache->list_table);
+        free(cache->map);
```
Move the statement `free(cache);` to the last.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,378 @@
+    cache->map = malloc(sizeof(hashtable_t));
+    if (!cache->map) {
+        free(cache);
+        free(cache->list_table);
```
Ditto. The order of `free` invocation.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,378 @@
+
+static inline void hlist_del_init(struct hlist_node *n)
+{
+    if (!hlist_unhashed(n)) {
```
Rewrite as following:
```c
    if (hlist_unhashed(n)
        return;
    hlist_del(n);
    INIT_HLIST_NODE(n);
```

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,378 @@
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include "cache.h"
```
Always ensure system headers prior to local ones. That is, move `cache.h` at the end of inclusion.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,378 @@
+    LFU_list,
+    LRU_ghost_list,
+    LFU_ghost_list,
+    list_quantity
```
Change the name `list_quantity` to `N_CACHE_LIST_TYPES` for better understanding.

jserv commented on tests/cache/cache.c:
```diff
@@ -0,0 +1,378 @@
+/*
```
Why did you duplicate `cache.c`?

jserv commented on mk/cache-test.mk:
```diff
@@ -0,0 +1,49 @@
+CACHE_TEST_DIR := tests/cache
+CACHE_BUILD_DIR := build/cache
+TARGET := test_cache.out
```
The naming scheme violates the consistency.
* `.out` is the suffix for the files generated by test programs. Typically, the text output.
* For host executables, there is no specific suffix.

Here, simply name it `test_cache`.

jserv commented on mk/cache-test.mk:
```diff
@@ -0,0 +1,49 @@
+	$(VECHO) "  CC\t$@\n"
+	$(Q)$(CC) $^ -o $(CACHE_BUILD_DIR)/$(TARGET)
+	
+$(CACHE_BUILD_DIR)/cache.o: $(CACHE_TEST_DIR)/cache.c
```
Don't duplicate `cache.c`. Always reuse existing files when possible.

jserv replied:
You shall honor the `CFLAGS` in top-level `Makefile`.

jserv commented on mk/cache-test.mk:
```diff
@@ -0,0 +1,49 @@
+	
+$(CACHE_BUILD_DIR)/cache.o: $(CACHE_TEST_DIR)/cache.c
+	$(VECHO) "  CC\t$@\n"
+	$(Q) mkdir -p $(CACHE_BUILD_DIR)
```
There is no space between `$(Q)` and `mkdir`.

jserv commented on src/common.h:
```diff
@@ -49,3 +49,28 @@
+    })
+#else
+#define container_of(ptr, type, member) \
+    ((type *) ((char *) (ptr) -offsetof(type, member)))
```
clang-format is sometimes stupid, and you have to write as following:
```c
#define container_of(ptr, type, member) \                                                                                                                     
    ((type *) ((char *) (ptr) - (offsetof(type, member))))
```

jserv commented on tests/cache/test_cache.c:
```diff
@@ -0,0 +1,63 @@
+    }
+}
+
+int main(int argc, char *argv[])
```
Add comments about the command sequences.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,388 @@
+    })
+#else
+#define hlist_entry_safe(ptr, type, member) \
+    ({ (ptr) ? hlist_entry(ptr, type, member) : NULL; })
```
The pair `({` and `})` is indeed a GNU extension, and we shall avoid using it for non-gcc/clang compilers. Later, rv32emu might support MSVC. Thus, we would write portable code as well.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,388 @@
+    return cache;
+}
+
+/* Rule of ARC
```
Should be "Rules"

jserv commented on mk/cache-test.mk:
```diff
@@ -0,0 +1,49 @@
+	
+$(CACHE_BUILD_DIR)/cache.o: src/cache.c 
+	$(VECHO) "  CC\t$@\n"
+	$(Q)mkdir -p $(CACHE_BUILD_DIR)
```
Replace this line with `$(Q)mkdir -p $(dir $@)`

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,388 @@
+
+void *cache_get(cache_t *cache, uint32_t key)
+{
+    if (cache->capacity <= 0 ||
```
The type of `capacity` is `uint32_t`, and it is always >= 0. Therefore, you shall never validate it with `<= 0`.

qwe661234 replied:
I modify it to 
``` c
if (!cache->capacity)
```

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,388 @@
+#define MIN(a, b) ((a < b) ? a : b)
+#define GOLDEN_RATIO_32 0x61C88647
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) % (cache_size)
```
We can specify the restriction for `cache_size`, which MUST be a power of 2. Then, `% (cache_size)` can be replaced with `& (cache_size - 1)`.

jserv commented on .github/workflows/main.yml:
```diff
@@ -15,6 +15,8 @@ jobs:
     - name: check
       run: make check
+    - name: cache
+      run: make cache
```
The target should not be `cache`. Instead, it would be `make tests`, which triggers the automated testing for all available tests. At the moment, cache specific items are associated.

qwe661234 replied:
Does this means changing the `name` column to `tests` while maintaining the `make cache` command in the `run` field? After this change, other tests would put command in the `run` field.
```
name: tests
run: make cache
     make XXX-test
     ...
```

jserv replied:
The name of GitHub Actions would accordingly be changed to `tests`.

jserv commented on mk/tests.mk:
```diff
@@ -0,0 +1,49 @@
+	$(VECHO) "  CC\t$@\n"
+	$(Q)$(CC) $^ -o $(CACHE_BUILD_DIR)/$(TARGET)
+	
+$(CACHE_BUILD_DIR)/cache.o: src/cache.c 
```
You can unify the rules by the following:
```
$(CACHE_BUILD_DIR)/%.o: $(CACHE_TEST_DIR)/%.c
```

qwe661234 replied:
The directory of `cache.c` is `src`, but the directory of `test_cache.c` is `tests/cache`. So I cannot nuify the rules to this.

jserv replied:
No, you don't have to build `$(CACHE_BUILD_DIR)/cache.o`. Instead, only `$(CACHE_BUILD_DIR)/test_cache.o` should be built here. The test oriented rules always depend on the generation of `$(OBJS)`.

qwe661234 replied:
I see, `cache.o` would be built by `make all`

qwe661234 replied:
However, the GitHub Action fails now, because we don't integrate the cache and it does not build `cache.o`.

jserv replied:
The appropriate sequence would be
```
make
make tests
```

Alternatively, you can extend item `check` to perform `make tests`.

jserv replied:
By the way, the next tests would be associated with #94 .

jserv commented on src/cache.h:
```diff
@@ -0,0 +1,40 @@
+ * @cache: a pointer points to target cache
+ * @release_entry: a function for freeing cache entry completely
+ */
+void cache_free(struct cache *cache, void (*release_entry)(void *));
```
Rename `release_entry` to `callback`.

jserv commented on mk/tests.mk:
```diff
@@ -0,0 +1,49 @@
+CACHE_BUILD_DIR := build/cache
+TARGET := test_cache
+
+CACHE_OBJS := cache.o \
```
Split `cache.o` into a new line.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,387 @@
+    n->pprev = &h->first;
+}
+
+static inline int hlist_unhashed(const struct hlist_node *h)
```
Change its type to `bool`.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,387 @@
+} hashtable_t;
+
+typedef struct cache {
+    struct list_head *list_table[N_CACHE_LIST_TYPES];
```
Rename `list_table` to `lists` for the sake of simplicity.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,387 @@
+    void *value;
+    uint32_t key;
+    cache_list_t arc_type;
+    struct list_head list;
```
Is it feasible to drop the use of `list` here?

qwe661234 replied:
`list` is for four cache lists T1, T2, B1, and B2 and `ht_list` is for hash table, we should maintain 4 lists by `list` and find out the certain cache entry by `ht_list`.

jserv replied:
Add corresponding comments along with further improvements.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,387 @@
+typedef struct {
+    void *value;
+    uint32_t key;
+    cache_list_t arc_type;
```
Rename `arc_type` to `type` since it is referred to a member of `cache_t` object.

jserv commented on src/cache.c:
```diff
@@ -0,0 +1,387 @@
+static uint32_t cache_size, cache_size_bits;
+
+/*
+ * ARC improves the basic LRU strategy by splitting the cache into two lists,
```
Refine the comments to reflect the meaning of each identifier enumerated.

jserv reviewed (changes requested):
For the consistency, file name would use dash (`-`) rather than underscores (`_`). One exception is to emphasize the items you want to validate. e.g., `test-cache_t`. In this pull request, we can safely enforce the replacement for all underscores.

jserv commented on .github/workflows/main.yml:
```diff
@@ -14,7 +14,9 @@ jobs:
     - name: default build
       run: make
     - name: check
```
This item should be `check + tests`.

jserv reviewed (changes requested):
Resolve CI pipeline failure.

qwe661234 committed:
Add adaptive replacement cache

Current basic block management consumes a significant amount of memory,
which leads to unnecessary waste due to frequent map allocation and
release. Adaptive Replacement Cache (ARC) is a page replacement
algorithm with better performance than least recently used (LRU). After
the translated blocks are handled by ARC, better memory usage and hit
rates can be achieved by keeping track of frequently used and recently
used pages, as well as a recent eviction history for both.

According to the cache information obtained while running CoreMark, the
cache hit rate of ARC can reach over 99%.

jserv commented on tests/cache/test-cache.c:
```diff
@@ -0,0 +1,69 @@
+    }
+}
+
+/* Commands of test_cache
```
Should be `test-cache`.

qwe661234 committed:
Add cache-test
